### ISSUE TITLE
Reflection possibly checks receiver type, but throws a different exception

### ISSUE THREAD

**Issue by jburse** [leaningtech/cheerpj-appletrunner#27]
I made some further testing, related to https://github.com/leaningtech/cheerpj-appletrunner/issues/26 . Was also testing how the reflection API deals with virtual calls. The finding is that the receiver type is possibly checked, but a different exception than usually is thrown. I started implementing get_byte/1 via a virtual call and found some discrepancies.

This is what Java usually gives:

![unbenannt](https://user-images.githubusercontent.com/5411487/50399634-2ab21400-0781-11e9-8dc5-60f2de7b1e28.PNG)

And this is what CheerpJ gives me:

![unbenannt2](https://user-images.githubusercontent.com/5411487/50399646-3ef61100-0781-11e9-92c0-11123368eaf9.PNG)

**Comment by jburse** [comment]
The matter can be analysed by looking at the exeception mapping we were using:

```
        } else if (x instanceof IllegalArgumentException) {
            return new EngineMessage(EngineMessage.permissionError(
                    AbstractFactory.OP_PERMISSION_APPLY,
                    mapMemberType(y),
                    mapMemberCulprit(y)));
        } else if (x instanceof NullPointerException) {
            return new EngineMessage(EngineMessage.permissionError(
                    AbstractFactory.OP_PERMISSION_LOOKUP,
                    mapMemberType(y),
                    mapMemberCulprit(y)));
```

So it seems that usual Java throws IllegalArgumentException, and CheerpJ throws NullPointerException. But the specification of method.invoke(obj, args); doesn't give me that NullPointerException is appropriate for receiver type mismatch:

> NullPointerException      if the specified object is null
> and the method is an instance method.

On the otherhand IllegalArgumentException is what the specification says:

> IllegalArgumentException  if the method is an
> instance method and the specified object argument
> is not an instance of the class or interface
> declaring the underlying method (or of a subclass
> or implementor thereof); if the number of actual
> and formal parameters differ; if an unwrapping
> conversion for primitive arguments fails; or if,
> after possible unwrapping, a parameter value
> cannot be converted to the corresponding formal
> parameter type by a method invocation conversion.

This is not a show stopper, since this doesn't render the reflection API unsafe. We mention this only for the record.

**Comment by alexp-sssup** [comment]
Fixed in CheerpJ Applet Runner 6.9

**Comment by jburse** [comment]
I did some regression testing. Does not yet work:

Ordinary Java:

obj=matula.util.system.ConnectionReader@49eee8b4
java.lang.IllegalArgumentException: object is not an instance of declaring class
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)

CheerpJ:

cheerpOS.js:1407 obj=matula.util.system.ConnectionReader@1891
cheerpOS.js:1407 java.lang.NullPointerException
cheerpOS.js:1407  	at java.io.Reader.read(Unknown Source)
cheerpOS.js:1407  	at sun.reflect.NativeMethodAccessorImpl.invoke0(Unknown Source)
cheerpOS.js:1407  	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
cheerpOS.js:1407  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
cheerpOS.js:1407  	at java.lang.reflect.Method.invoke(Unknown Source)

**Comment by alexp-sssup** [comment]
We have no way of debugging the problem without a reproducible test case. Can you create one?

**Comment by jburse** [comment]
Please give me some time. I first try to get an overview what can be closed 
and what cannot yet be closed, regarding reflection. I am also having a more 
closer look than before. I will try to write down a Java test case.

**Comment by jburse** [comment]
Here is a test case for this issue:

1. Test Program:

```
01 public class Test {
02
03    public static void main(String[] args) {
04        try {
05            Reader reader = new StringReader("abc");
06            Method method = InputStream.class.getDeclaredMethod("available");
07            method.invoke(reader);
08        } catch (Exception x) {
09            x.printStackTrace();
10        }
11    }
12 }
```
2. Expected Result, from ordinary Java, some exception:

```
java.lang.IllegalArgumentException: object is not an instance of declaring class
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at Test.test(Test.java:7)
```
See also: Throws clause here:
https://docs.oracle.com/javase/7/docs/api/java/lang/reflect/Method.html#invoke(java.lang.Object,%20java.lang.Object...)

3. Actual Result from CheerpJ, no exception

![unbenannt4](https://user-images.githubusercontent.com/5411487/51416881-c44fc400-1b7b-11e9-9e1e-26e55a299f3b.PNG)

**Comment by alexp-sssup** [comment]
Fixed in CheerpJ Applet Runner 6.10
